packages/flutter/lib/src/widgets/router.dart

```diff
@@ -998,6 +998,76 @@ class ChildBackButtonDispatcher extends BackButtonDispatcher {
   }
 }
 
+/// A convenience widget that registers a callback when the back button is pressed
+/// by using [BackButtonDispatcher].
+///
+/// It's required to have a [Router] widget ancestor on the tree.
+///
+/// It only applies to platforms that accept back button clicks, such as Android.
+///
+/// It can be useful for scenarios, in which you create a different state in your
+/// screen but don't want to use a new page for that.
+class BackButtonListener extends StatefulWidget {
+  /// Creates a BackButtonListener widget .
+  ///
+  /// The [child] and [onBackPressed] arguments must not be null.
+  const BackButtonListener({
+    Key? key,
+    required this.child,
+    required this.onBackPressed,
+  }) : super(key: key);
+
+  /// The [child] Widget contained by the BackButtonListener.
+  final Widget child;
+
+  /// The callback function that will be called when the back button is pressed.
+  ///
+  /// It must return a boolean future with true if this child will handle the request;
+  /// otherwise, return a boolean future with false.
+  final ValueGetter<Future<bool>> onBackPressed;
+
+  @override
+  _BackButtonListenerState createState() => _BackButtonListenerState();
+}
+
+class _BackButtonListenerState extends State<BackButtonListener> {
+  BackButtonDispatcher? dispatcher;
```

Comment:
I think this can be
```dart
late BackButtonDispatcher dispatcher;
```

Reply:
I believe this would not work well, because the first thing we need to do on `didChangeDependencies` is check if the `dispatcher` is null or not to remove the previous callback. WDYT?

Reply:
We could otherwise move the `assert(rootBackDispatcher != null)` to the top of this method and consider it an invalid state otherwise

Reply:
you are right, i missed that part